### PR TITLE
.Net adding chat completion and settings to Oobabooga connector

### PR DESCRIPTION
**Pull request by jsboige**
### Motivation and Context

Resolves: #2583 

a Oobabooga text completion connector was recently added to SK. Oobabooga API also supports a rich Chat API compatible with SK's chat completion model

### Description

This PR completes the recently added Oobabooga connector with a complementary chat completion connector and exposes Oobabooga settings both for text and chat completion.

### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/OobaboogaCompletionBase.cs [hunk 1]
```diff
@@ -2,43 +2,30 @@
...
-    private readonly HttpClient _httpClient;
-    private readonly Func<ClientWebSocket> _webSocketFactory;
-    private readonly bool _useWebSocketsPooling;
-    private readonly int _maxNbConcurrentWebSockets;
-    private readonly SemaphoreSlim? _concurrentSemaphore;
-    private readonly ConcurrentBag<bool>? _activeConnections;
-    private readonly ConcurrentBag<ClientWebSocket> _webSocketPool = new();
-    private readonly int _keepAliveWebSocketsDuration;
-    private readonly ILogger? _logger;
+    public const string HttpUserAgent = "Microsoft-Semantic-Kernel";
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/OobaboogaCompletionBase.cs:18]: We use `Telemetry.HttpUserAgent` from `Microsoft.SemanticKernel.Diagnostics` to align user-agent value across all connectors.

**Review by dmytrostruk — COMMENTED** [review]
@jsboige Thank you for contribution! I left initial feedback. In order to proceed with this PR, please create GitHub Issue. Our team will triage and prioritize it accordingly. Thanks again!

File: dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/ChatCompletionOobaboogaSettings.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    /// The character name for the chat completion.
+    /// </summary>
+    [JsonPropertyName("character")]
+    public string Character { get; set; } = "Example";
+
+    /// <summary>
+    /// The instruction template for the chat completion.
+    /// </summary>
+    [JsonPropertyName("instruction_template")]
+    public string InstructionTemplate { get; set; } = "Vicuna-v1.1";
```
[the remark this answers is not in the document]
Comment (jsboige) [reply dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/ChatCompletionOobaboogaSettings.cs]: I added documentation and external references in [this commit](https://github.com/microsoft/semantic-kernel/pull/2229/commits/78479ec41f98d57a93733501f87780a6e75719c1#diff-9c57c6bdb3c7324ba3bf2cebffe131b7bc4fc444cc99d12e084ffbae68cfc34cR26)

File: dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/ChatCompletionOobaboogaSettings.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    public string ChatInstructCommand { get; set; } = "Continue the chat dialogue below. Write a single reply for the character \"<|character|>\".\n\n<|prompt|>";
+
+    /// <summary>
+    /// The instruction context for the chat-instruct / instruct completion.
+    /// </summary>
+    [JsonPropertyName("context_instruct")]
+    public string ContextInstruct { get; set; } = "";
+
+    /// <summary>
+    /// Imports the settings from the given ChatCompletionRequestSettings object.
```
[the remark this answers is not in the document]
Comment (jsboige) [reply dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/ChatCompletionOobaboogaSettings.cs]: Done in [this commit](https://github.com/microsoft/semantic-kernel/pull/2229/commits/8d50aff0b5ac7b78cfce3ef103aa5e3e1d7da41c)

File: dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/ChatCompletionRequest.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    /// <summary>
+    /// The chat history.
+    /// </summary>
+    [JsonPropertyName("history")]
+    public ChatHistory History { get; set; } = new ChatHistory();
+
+    /// <summary>
+    /// Creates a new ChatCompletionRequest with the given Chat history, oobabooga settings and semantic-kernel settings.
+    /// </summary>
+    public static ChatCompletionRequest Create(SemanticKernel.AI.ChatCompletion.ChatHistory chat, ChatCompletionOobaboogaSettings settings, ChatRequestSettings requestSettings)
```
[the remark this answers is not in the document]
Comment (jsboige) [reply dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/ChatCompletionRequest.cs]: Done in [this commit](https://github.com/microsoft/semantic-kernel/pull/2229/commits/92befb19a6ec7173a78d682abce7b626160bf98f)

File: dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/ChatCompletionRequest.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        var chatMessages = chat.Messages.Take(chat.Messages.Count - 1).Select(@base => @base.Content).ToList();
+        var toReturn = new ChatCompletionRequest()
+        {
+            UserInput = chat.Messages.Last().Content,
```
[the remark this answers is not in the document]
Comment (jsboige) [reply dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/ChatCompletionRequest.cs]: Done in [this commit](https://github.com/microsoft/semantic-kernel/pull/2229/commits/83e3e30a33bbdb21e7991aea40fb036a3c575114)

File: dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/ChatCompletionResult.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    public ChatCompletionResult(ChatCompletionResponseHistory responseData)
+    {
+        this._responseData = new ModelResult(responseData);
+    }
+
+    public ModelResult ModelResult => this._responseData;
+
+    public Task<string> GetCompletionAsync(CancellationToken cancellationToken = default)
+    {
+        return Task.FromResult(this._responseData.GetResult<TextCompletionResponseText>().Text ?? string.Empty);
```
[the remark this answers is not in the document]
Comment (jsboige) [reply dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/ChatCompletionResult.cs]: It was leftover code that was [removed](https://github.com/microsoft/semantic-kernel/pull/2229/commits/2691b602cb3375472ec52ef5d20f4fde27755b23)

File: dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/OobaboogaChatCompletion.cs [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+
+            using var httpRequestMessage = new HttpRequestMessage()
+            {
+                Method = HttpMethod.Post,
+                RequestUri = this._chatBlockingUri.Uri,
+                Content = stringContent
+            };
+            httpRequestMessage.Headers.Add("User-Agent", HttpUserAgent);
+
+            using var response = await this.HTTPClient.SendAsync(httpRequestMessage, cancellationToken).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (jsboige) [reply dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/OobaboogaChatCompletion.cs]: Done

File: dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/OobaboogaChatCompletion.cs [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+                AIException.ErrorCodes.UnknownError,
+                $"Something went wrong: {e.Message}", e);
+        }
+        finally
+        {
+            this.FinishConcurrentCall();
+        }
+    }
+
+    public async IAsyncEnumerable<IChatStreamingResult> GetStreamingChatCompletionsAsync(
```
[the remark this answers is not in the document]
Comment (jsboige) [reply dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/OobaboogaChatCompletion.cs:55]: done

[the remark this answers is not in the document]
Comment (jsboige) [reply dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/OobaboogaChatCompletion.cs:55]: Done in the large refactoring comprising my latest commits. That involved introducing an intermediate class with many generic parameters to account for all the component classes differing between Text and Chat completion. That many generics isn't super pretty, but it is an intermediate class in the inheritance hierarchy, so it shouldn't bother the end users.

File: dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/OobaboogaCompletionBase.cs (hunk 1 above)
Comment (jsboige) [reply dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/OobaboogaCompletionBase.cs]: Done

File: dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/OobaboogaCompletionBase.cs [hunk 9]
```diff
@@ -2,43 +2,30 @@
...
+
+    internal HttpClient HTTPClient;
+    internal Func<ClientWebSocket> WebSocketFactory;
+    internal bool UseWebSocketsPooling;
```
[the remark this answers is not in the document]
Comment (jsboige) [reply dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/OobaboogaCompletionBase.cs]: Done (well, done initially, though now a lot has moved in the latest refactoring, I'll come back to this below)

File: dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/ChatCompletionResult.cs [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    public Task<ChatMessageBase> GetChatMessageAsync(CancellationToken cancellationToken = default)
+    {
+        return Task.FromResult((ChatMessageBase)new SKChatMessage(this._responseData.GetResult<ChatCompletionResponseHistory>().History.Visible.Last()));
```
[the remark this answers is not in the document]
Comment (jsboige) [reply dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/ChatCompletionResult.cs]: Done

File: dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/OobaboogaChatCompletion.cs [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.SemanticKernel.AI;
+using Microsoft.SemanticKernel.AI.ChatCompletion;
+using Microsoft.SemanticKernel.Diagnostics;
+
+namespace Microsoft.SemanticKernel.Connectors.AI.Oobabooga.Completion.ChatCompletion;
+
+/// <summary>
+/// Oobabooga text completion service API.
```
[the remark this answers is not in the document]
Comment (jsboige) [reply dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/OobaboogaChatCompletion.cs]: Done

File: dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/OobaboogaChatCompletion.cs [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+/// Adapted from <see href="https://github.com/oobabooga/text-generation-webui/tree/main/api-examples"/>
+/// </summary>
+public sealed class OobaboogaChatCompletion : OobaboogaCompletionBase, IChatCompletion
```
[the remark this answers is not in the document]
Comment (jsboige) [reply dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/OobaboogaChatCompletion.cs]: Done

File: dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/OobaboogaChatCompletion.cs [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+    private readonly UriBuilder _chatBlockingUri;
+    private readonly UriBuilder _chatStreamingUri;
+
+    public const string ChatBlockingUriPath = "/api/v1/chat";
+    public const string ChatStreamingUriPath = "/api/v1/chat-stream";
+
+    public ChatCompletionOobaboogaSettings ChatCompletionOobaboogaSettings { get; set; }
+
+    /// <inheritdoc/>
+    public OobaboogaChatCompletion(Uri endpoint, int blockingPort = 5000,
```
[the remark this answers is not in the document]
Comment (jsboige) [reply dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/OobaboogaChatCompletion.cs]: Done

[the remark this answers is not in the document]
Comment (jsboige) [reply dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/OobaboogaChatCompletion.cs]: OK this triggered the largest of my refactoring changes over several commits. I'll make a general comment below

File: dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/OobaboogaChatCompletion.cs [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+/// Oobabooga text completion service API.
+/// Adapted from <see href="https://github.com/oobabooga/text-generation-webui/tree/main/api-examples"/>
+/// </summary>
+public sealed class OobaboogaChatCompletion : OobaboogaCompletionBase, IChatCompletion
+{
+    private readonly UriBuilder _chatBlockingUri;
+    private readonly UriBuilder _chatStreamingUri;
+
+    public const string ChatBlockingUriPath = "/api/v1/chat";
+    public const string ChatStreamingUriPath = "/api/v1/chat-stream";
```
[the remark this answers is not in the document]
Comment (jsboige) [reply dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/OobaboogaChatCompletion.cs]: Done

File: dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/OobaboogaChatCompletion.cs [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+
+    public ChatCompletionOobaboogaSettings ChatCompletionOobaboogaSettings { get; set; }
```
[the remark this answers is not in the document]
Comment (jsboige) [reply dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/OobaboogaChatCompletion.cs]: Done

File: dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/OobaboogaChatCompletion.cs [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+            Port = streamingPort,
+            Path = ChatStreamingUriPath
+        };
+        if (this._chatStreamingUri.Uri.Scheme.StartsWith("http", StringComparison.OrdinalIgnoreCase))
+        {
+            this._chatStreamingUri.Scheme = this._chatStreamingUri.Scheme == "https" ? "wss" : "ws";
+        }
+    }
+
+    public SemanticKernel.AI.ChatCompletion.ChatHistory CreateNewChat(string? instructions = null)
```
[the remark this answers is not in the document]
Comment (jsboige) [reply dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/OobaboogaChatCompletion.cs]: Done

File: dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/OobaboogaChatCompletion.cs [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+    {
+        return new SemanticKernel.AI.ChatCompletion.ChatHistory();
+    }
+
+    public async Task<IReadOnlyList<IChatResult>> GetChatCompletionsAsync(
```
[the remark this answers is not in the document]
Comment (jsboige) [reply dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/OobaboogaChatCompletion.cs:45]: Done

File: dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/OobaboogaChatCompletion.cs [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+                JsonSerializer.Serialize(completionRequest),
+                Encoding.UTF8,
+                "application/json");
+
+            using var httpRequestMessage = new HttpRequestMessage()
+            {
+                Method = HttpMethod.Post,
+                RequestUri = this._chatBlockingUri.Uri,
+                Content = stringContent
+            };
```
[the remark this answers is not in the document]
Comment (jsboige) [reply dotnet/src/Connectors/Connectors.AI.Oobabooga/Completion/ChatCompletion/OobaboogaChatCompletion.cs]: Done

**Comment by jsboige** [comment]
@dmytrostruk Thanks for your review.
I believe I addressed all of your comments and concerns.
From my commits since yesterday, the first address your comments individually, but then 2 issues involved a major refactoring with major changes:

- My initial PR included an initial effort at mutualizing some code between chat and text completion. I had also noted that I could have gone further as you did, but that was a bit tricky since the code was similar but making use of a series of distinct classes for request, response, results etc. Accordingly, that involved starting making use of generics to support Text/Chat with all those differences.

- Then the major refactoring pushed that further to get rid of the many ctor parameters by moving a lot of the logic into the Settings classes, which used to inherit from the Oobabooga's parameters, and now hold them as an inner property, additionally to the parameters that used to be within the completion classes.
Accordingly, there is a several step hierarchy through generics for settings as well as for completion.

**Comment by jsboige** [comment]
> @jsboige Thank you for contribution! I left initial feedback. In order to proceed with this PR, please create GitHub Issue. Our team will triage and prioritize it accordingly. Thanks again!

Here is the [corresponding issue](https://github.com/microsoft/semantic-kernel/issues/2583)

**Comment by dmytrostruk** [comment]
@jsboige Thanks for all the hard work! 
Before performing next round of review, I would like to share my general feedback and our long-term vision. It would be great to know your thoughts. 

In general, as this functionality grows, it looks like separate Oobabooga .NET SDK, while in this repository we are focused more on lightweight connectors, that can be used with Semantic Kernel. Ideal scenario is when integration client is implemented as separate package, and we re-use it when implementing `ITextCompletion` and `IChatCompletion` interfaces (like adapters).

At the moment we are looking at opportunity to extract connectors and plugins to separate repositories (PR with contribution guidelines: https://github.com/microsoft/semantic-kernel/pull/2503). We are working on mechanisms and will post official guidelines soon.

Meanwhile, do you think it will be possible for you to migrate Oobabooga client implementation to separate repository and package, so in this connector we will reference it and use necessary methods to implement `ITextCompletion` and `IChatCompletion` interfaces, so it can be used with Semantic Kernel? 

Please let me know what you think about it, thank you!

**Comment by jsboige** [comment]
> @jsboige Thanks for all the hard work! Before performing next round of review, I would like to share my general feedback and our long-term vision. It would be great to know your thoughts.
> 
> In general, as this functionality grows, it looks like separate Oobabooga .NET SDK, while in this repository we are focused more on lightweight connectors, that can be used with Semantic Kernel. Ideal scenario is when integration client is implemented as separate package, and we re-use it when implementing `ITextCompletion` and `IChatCompletion` interfaces (like adapters).
> 
> At the moment we are looking at opportunity to extract connectors and plugins to separate repositories (PR with contribution guidelines: #2503). We are working on mechanisms and will post official guidelines soon.
> 
> Meanwhile, do you think it will be possible for you to migrate Oobabooga client implementation to separate repository and package, so in this connector we will reference it and use necessary methods to implement `ITextCompletion` and `IChatCompletion` interfaces, so it can be used with Semantic Kernel?
> 
> Please let me know what you think about it, thank you!

@dmytrostruk Thanks for following through with reviewing yet another PR of mine, and for extending your long-term vision.

Regarding #2503, I understand the motivation, which makes sense and, similarly to what happened with the recent extraction of chat-copilot, I'll be happy to commit to another dedicated connector's repository to keep this one light and focused.

Then extending the idea to extracting the Oobabooga's client from the connector itself, the way I understand it, you would like to reproduce the existing architecture of the OpenAI connector, which is implemented through as a wrapper to [Azure SDK's Open AI client library](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/openai/Azure.AI.OpenAI#azure-openai-client-library-for-net)'s Nugget package, that library being an external projet that can be used to target Open AI models independently from semantic-kernel.  Again that makes sense, and I suppose most connectors and plugin are meant to wrap some kind of external client library with bindings to the core interfaces and abstractions.

Here are several considerations that come to mind:

- My company is a small single man's operation, my main occupation at the moment is teaching, and that side project of a [card's game about argumentation](https://github.com/ArgumentumGames/Argumentum) is probably going to need a significant effort of mine in the near term even though I'm definitely planning to make SK a part of that project through the plugin I started to draft during the hackathon. Accordingly, taking ownership of an external repository is an extra commitment I should not take lightly, but I understand that also as an opportunity for me to grow my business by becoming a close partner to the official semantic-kernel channels, which I believe I'm willing to do provided I can make it somehow sustainable.

- Then concerning the opportunity of extracting an external independent Oobabooga connector, I am not sure there is yet the material to do so.
   - Looking at the code itself, there does not seem to be much to extract yet from the SK's wrapper. Just like the Http client logic is very similar to Huggingface's and I could see opportunities to try and abstract some back into the kernel, the websocket logic from the streaming endpoint also looks pretty general to me, and that would leave us with just about the couple classes in charge of implementing the specific json text and chat request and response serialization schemas.
   
   - Now if you start adding embeddings, image generations etc. that would draw a different picture, but the way oobabooga is currently evolving, they chose to implement those through a [distinct extension API](https://github.com/oobabooga/text-generation-webui/blob/main/extensions/openai/README.md) that emulates Open AI's and that would actually perfectly fit adding a third implementation to the existing OpenAI's connector beside the Azure and OpenAI existing ones, because the needed abstraction work was already done for those 2; it should be straightforward to implement it on top of the [same base classes](https://github.com/microsoft/semantic-kernel/tree/main/dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk) and while I don't see myself adding much to this specific oobabooga connector in the near future as I believe the main official API is now entirely covered, the OpenAI extension is actually one of the next contributions I'm planning to bring to SK, whether it's inside the actual project or referencing the [corresponding Nugget package](https://www.nuget.org/packages/Microsoft.SemanticKernel.Connectors.AI.OpenAI/) (according to your preferences) in a distinct repository. 
   
   - Then I also wonder if anyone would be willing to use such a connector outside of semantic-kernel. My understanding of the market is that most oobabooga users are Python developers, which I'm not, and they won't be willing to use a c# connector unless it brings them something they don't already find in Langchain or Guidance, typically what SK brings, either through Python.Net or a chat-copilot equivalent using external plugins of theirs, but I suppose preferably through a Python port of the connector. Then some .Net developers might be interested in a raw oobabooga connector, but to be honest I'm not, and I doubt anyone will be willing to push for that in Microsoft's sphere as it might be perceived as interfering with the standard toolkit more than through this SK connector.
   
   - This is not to say I don't see myself doing that extraction into yet another repository, but I'd rather keep that for a later term if you don't mind.
   
- I suppose the same consideration applies to the [multiconnector](https://github.com/microsoft/semantic-kernel/pull/2323) I recently introduced. If any, that one surely has the potential to become a major plugin and would need it's own external repository soon, as apparently the [demand for such a component is big](https://twitter.com/Altimor/status/1696221744024142210). I believe there is a natural continuation of that connector leveraging [Infer.Net](https://github.com/dotnet/infer), which I'm willing to implement, based on a probabilistic model accommodating those 4 examples:
   - [Student Skills](https://dotnet.github.io/infer/userguide/Student%20skills.html)
   - [Assessing People's Skills](https://mbmlbook.com/LearningSkills.html)
   - [Difficulty vs Ability](https://dotnet.github.io/infer/userguide/Difficulty%20versus%20ability.html)
   - [Calibrating reviews](https://dotnet.github.io/infer/userguide/Calibrating%20reviews%20of%20conference%20submissions.html) 

- Add to that that I plan to implement a complementary distributed [Spark.Net](https://github.com/dotnet/spark) based connector, which I believe should be relatively straight-forward too, and that starts to make for a serious semantic-kernel code base to maintain.  

Now here's for my side of the story.  
I'm now teaching AI in .Net, after a previous career where I was a DNN evangelist in France, with several open source extensions including [one that I'd like to revive with SK](https://github.com/Aricie/Aricie.PortalKeeper), that did not meet their public. 

Doing and teaching AI in .Net has proved exhausting with most projects ending up losing traction one way or another (Solver foundation dead, Encog dead, Accord.Net too, Genetic Sharp stalled, AIMA.Net dead, DlxLib dead, Quickgraph dead, Heuristiclab dead, CNTK dead, KerasSharp dead, TensorflowSharp dead, etc.). Even those still active like Z3, OR-tools, ML.Net, Spark.Net, SciSharp, TorchSharp, Infer etc. are not super healthy. I hope you guys are aware of that grim history.

With that said I'm still pretty committed to that .Net stack and I'd rather see IronPython and IKVM properly revived that take the Python.Net bridging path, which I started doing occasionally.
This is to say I really like this library, I love what LLMs have to bring, and I see this project as one of the latest chance for c# to shine in the AI landscape, which probably is the reason I've been willing to invest so much in it over the last couple months. 

Also I tried to avoid past pitfalls where significant contributions of mine would be wasted like [this one](https://github.com/giacomelli/GeneticSharp/pull/87) or [that one](https://github.com/RicardoNiepel/Z3.LinqBinding/pull/2) although I just realized the latest was eventually partially merged [into a revival project](https://github.com/endjin/Z3.Linq/issues/29). 

Accordingly, I'm super grateful my contributions were accepted with open arms, and I believe I'm now ready to become the steward of some of it if you think that's the best path forward. 
Let me know how best to proceed and I'll try to make some time to accommodate your decisions.

**Comment by dmytrostruk** [comment]
@jsboige Thank you very much for your response and great contribution to this repository!
We will provide the next steps as soon as we align with the process. Let's keep this PR open for now.
Thanks again!

**Comment by nacharya1** [comment]
@jsboige  FYI: We recently updated our contributing guidelines [here](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#adding-plugins-and-memory-connectors , stating we are not directly adding plugin and connectors directly into the SDK repository. We are encouraging the community to host their own plugins and connectors separately in their own repositories. At the time we are exploring how we can provide a way to discover plugins and connectors and should have an update soon. I am going to close this issue given our change in contribution guidelines. 
cc: @dmytrostruk

**Comment by jsboige** [comment]
@nacharya1, @dmytrostruk 
OK I will start building an alternate repo. I will let you know about it.

**Pull request closed without merging**